arraycontext/context.py

```diff
@@ -123,10 +113,55 @@
 from pytools.tag import Tag
 
 
-DeviceArray = Any
-DeviceScalar = Any
+# {{{ typing
+
 _ScalarLike = Union[int, float, complex, np.generic]
 
+try:
+    from typing import Protocol
+except ImportError:
+    from typing_extensions import Protocol                  # type: ignore[misc]
+
+
+class DeviceArray(Protocol):
```

Comment:
Maybe use this in the type annotations for the containers? (That's undoubtedly a bigger project, so we can alway tackle that later.)

Reply:
I think most of the code in `arraycontext.container.traversal` already uses it a bit. I have another branch/PR on top of this that adds more annotations to `ArrayContext` using these, so we can give it a more thorough look there.